eyetrackprep/src/pupil2bids.py

```diff
@@ -305,6 +303,333 @@ def compile_rawfile_list(
+        """
+        If the log file cannot be parsed, estimate the scanner onset time 
+        from the run's 12th gaze timestamp (estimated from differences between 
+        gaze timestamps and logged TTL 0 for 14 CNeuroMod tasks with eye-tracking data)
+        """        
+        log_qc(f"Warning: failed log parsing, run onset and offset time estimated from gaze timestamps for {fnum}", qc_path)
+        return gz0_ts, gzn_ts, "n/a"
+
+
+def extract_gaze(
+    seri_gaze,
+    onset_time,
+) -> list:
+    """
+    Converts nested gaze dictionary to two lists of arrays (one for BIDS, one for plotting)
+    """
+    bids_gaze_list = []
+
+    for gaze in seri_gaze:
+        gaze_timestamp = gaze['timestamp'] - onset_time
+        if gaze_timestamp > 0.0:
+            gaze_x, gaze_y = gaze['norm_pos']
+            gaze_conf = gaze['confidence']
+            pupil_data = gaze['base_data'][0]
+            pupil_x, pupil_y = pupil_data['norm_pos']
+            pupil_diameter = pupil_data['diameter']
+            ellipse_data = pupil_data['ellipse']
+            ellipse_axe_a, ellipse_axe_b = ellipse_data['axes']
+            ellipse_angle = ellipse_data['angle']
+            ellipse_center_x, ellipse_center_y = ellipse_data['center']
+
+            bids_gaze_list.append([
+                gaze_timestamp, gaze_x, gaze_y, gaze_conf, 
+                pupil_x, pupil_y, pupil_diameter, 
+                ellipse_axe_a, ellipse_axe_b, ellipse_angle, 
+                ellipse_center_x, ellipse_center_y,
+            ]) 
+                
+    return bids_gaze_list
+
+
+def detect_freezes(
+    gaze: np.array,
+    out_file: str,
+    run_duration: float,
+) -> float:
+    """
+    Identifies temporal gaps in eye-tracking data due to camera freezes, 
+    and exports them as physioevents.tsv file.
+
+    Detects instances of inter-sample interval > 0.005s (indicating a dropped frame 
+    or camera freeze, based on a 250Hz pupil sampling rate). 
+
+    Args:
+        gaze (np.array): A 2D array of gaze data with timestamps in col 0 (in s). 
+        out_file (str): The output file name for a particular run.
+        run_duration (float): The estimated run duration (in s).
+
+    Returns:
+        int: number of camera freezes
+
+    If camera freezes are detected, two files are exported:
+    - `physioevents.tsv.gz`: A compressed tab-separated file with 'onset' and 
+        'duration' for the identified camera freezes.
+    - `events.json`: A metadata file describing the columns in the TSV.
+    """
+    ts_arr = np.stack((
+            gaze[:-1, 0], gaze[1:, 0] - gaze[:-1, 0],
+        ), axis=1,
+    )
+    #ts_arr = ts_arr[ts_arr[:, 1] > 0.005]  # too sensitive: detects any skipped frame...
+    ts_arr = ts_arr[ts_arr[:, 1] > 0.5]  # > 0.5s freezes only
+
+    if gaze[0, 0] > 1.0:  
+        ts_arr = np.concat([
+            np.array([[0.0, gaze[0, 0]]]),
+            ts_arr,
+        ], axis=0)
+        
+    if run_duration - gaze[-1, 0] > 2.0:
```

Comment:
Where is the `2.0` coming from ? Possible to add a comment either here or in the function description ?